storage/innobase/include/srv0tmp.h

```diff
@@ -176,7 +176,24 @@ class Tablespace_pool {
     release();
   }
 
+  THD *get_thd_for_ts(space_id_t id) const noexcept {
+    /* No lock needed:
+
+       the map used is only modified in expand(), where first a new map is
```

Comment:
I don't think this works either. innodb_temp_table_thds is a unique ptr. When expand() is called, the old one is deallocated. What happens if it gets deallocated while the find is running on the old map?

It seems this might be difficult to do without acquiring a lock to a global map in the I/O threads. @lth who worked on the original change has a suggestion for returning the size of the temp_table at the end of the query vs. returning bytes written. This is mainly a proxy for detecting slow queries that are trying to create large temp table files. If it's easier for the thd to pull out the size of the filespace after the query and record that in the session variable, we can do that too.

Thoughts?

Reply:
That would work, as that information is already available and used in information_schema, could be easily added to any session variable  - but it would be a different number. Shouldn't it be a different variable then?

Or I still think this can be fixed - I wanted to simplify it with a single unique ptr and a map, but I missed this problem. I also missed that tablespace ids here are increasing from a reserved range, so they can't be any value, which simplifies things further.

We know the possible maximum size of this structure (the maximum of max_connections, or the size of the reserved tablespace range), so instead of a map, I could use a fixed std::array, which then can be modified without issues. The downside is fixed memory usage, which is quite high for servers with a few connections (1600KB). We could do a two-step version, let's say 1000 and max, only allocating the larger structure when we reach 1000 connections, and never free any. (There are some other structures in the codebase that do this, for the same reason. a few megabytes doesn't really matter if you already have more than 1000 connections)

(tablespace ids start with a fixed id and then increase by one. But we might have holes, as we only care about user tables).

Reply:
With connection pooling, servers don't usually have many connections. So allocating a large map to track the space_id usage is probably not great. There are also possibly many small use cases of mysql that use temp tables, so a few megabytes of memory adds up.

Let's go with the size idea. We could record the before/after sizes of the temp tables and use that as "bytes" written. We lose the bytes written on updates, but that would probably be fine for now.